src/detect-mqtt-reason-code.c

```diff
@@ -0,0 +1,318 @@
+void DetectMQTTReasonCodeFree(DetectEngineCtx *de_ctx, void *);
+
+static int DetectEngineInspectMQTTReasonCodeGeneric(ThreadVars *tv,
+        DetectEngineCtx *de_ctx, DetectEngineThreadCtx *det_ctx,
+        const Signature *s, const SigMatchData *smd,
+        Flow *f, uint8_t flags, void *alstate,
+        void *txv, uint64_t tx_id);
+
+/**
+ * \brief Registration function for mqtt.reason_code: keyword
+ */
+void DetectMQTTReasonCodeRegister (void)
+{
+    sigmatch_table[DETECT_AL_MQTT_REASON_CODE].name = "mqtt.reason_code";
+    sigmatch_table[DETECT_AL_MQTT_REASON_CODE].alias = "mqtt.connack.return_code";
+    sigmatch_table[DETECT_AL_MQTT_REASON_CODE].desc = "match MQTT 5.0+ reason code";
+    sigmatch_table[DETECT_AL_MQTT_REASON_CODE].url = "/rules/mqtt-keywords.html#mqtt-reason-code";
+    sigmatch_table[DETECT_AL_MQTT_REASON_CODE].AppLayerTxMatch = DetectMQTTReasonCodeMatch;
+    sigmatch_table[DETECT_AL_MQTT_REASON_CODE].Setup = DetectMQTTReasonCodeSetup;
+    sigmatch_table[DETECT_AL_MQTT_REASON_CODE].Free  = DetectMQTTReasonCodeFree;
+    sigmatch_table[DETECT_AL_MQTT_REASON_CODE].RegisterTests = MQTTReasonCodeRegisterTests;
+
+    DetectSetupParseRegexes(PARSE_REGEX, &parse_regex);
+
+    DetectAppLayerInspectEngineRegister("mqtt.reason_code",
+            ALPROTO_MQTT, SIG_FLAG_TOSERVER, 1,
+            DetectEngineInspectMQTTReasonCodeGeneric);
+
+    mqtt_reason_code_id = DetectBufferTypeGetByName("mqtt.reason_code");
+}
+
+static int DetectEngineInspectMQTTReasonCodeGeneric(ThreadVars *tv,
+        DetectEngineCtx *de_ctx, DetectEngineThreadCtx *det_ctx,
+        const Signature *s, const SigMatchData *smd,
+        Flow *f, uint8_t flags, void *alstate,
+        void *txv, uint64_t tx_id)
+{
+    return DetectEngineInspectGenericList(tv, de_ctx, det_ctx, s, smd,
+                                          f, flags, alstate, txv, tx_id);
+}
+
+/**
+ * \internal
+ * \brief Function to match reason code of an MQTT 5.0 Tx
+ *
+ * \param det_ctx Pointer to the pattern matcher thread.
+ * \param f       Pointer to the current flow.
+ * \param flags   Flags.
+ * \param state   App layer state.
+ * \param txv     Pointer to the transaction.
+ * \param s       Pointer to the Signature.
+ * \param ctx     Pointer to the sigmatch that we will cast into DetectMQTTReasonCodeData.
+ *
+ * \retval 0 no match.
+ * \retval 1 match.
+ */
+static int DetectMQTTReasonCodeMatch(DetectEngineThreadCtx *det_ctx,
+                               Flow *f, uint8_t flags, void *state,
+                               void *txv, const Signature *s,
+                               const SigMatchCtx *ctx)
+{
+    const uint8_t *de = (const uint8_t *)ctx;
+    uint8_t code;
+
+    if (!de)
+        return 0;
+
+    if (rs_mqtt_tx_get_reason_code(txv, &code) == 0) {
+        bool result = false;
+        rs_mqtt_tx_unsuback_has_reason_code(txv, &result, *de);
+        return (result ? 1 : 0);
+    } else {
+        if (code == *de)
+            return 1;
```

Comment:
No, because `rs_mqtt_tx_unsuback_has_reason_code` does not return a code that needs to be compared. In that case, we directly get the result (match/no match) from Rust that we can directly use.
The reason here is that UNSUBSCRIBE can have multiple reason codes and I did not want to pass around multiple values between C and Rust, so the check is implemented in Rust.
I will add a comment to explain this.

Reply:
Thanks